### PR TITLE
MPWinogradWRW solver update

### PR DESCRIPTION
**Pull request by shurale-nkn**
MultipassWinogradWrW3x{4,5,6} disabled by default for fp16 with stride=1 on gfx908

### CODE REVIEW

File: src/solver/conv_multipass_wino3x3WrW.cpp [hunk 1]
```diff
@@ -381,15 +381,33 @@ bool ConvWinograd3x3MultipassWrW<WinoDataH, WinoFilterH, WinoDataW, WinoFilterW>
...
-            return false;
-    if(wino_data_tile == 3 && wino_filter_tile == 5)
-        if(miopen::IsDisabled(MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_F3X5{}))
-            return false;
-    if(wino_data_tile == 3 && wino_filter_tile == 6)
-        if(miopen::IsDisabled(MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_F3X6{}))
-            return false;
+
+    const std::string name = params.GetStream().GetDeviceName();
+    if(params.IsFp16() && StartsWith(name, "gfx908") && params.kernel_stride_h == 1)
```
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:386]: I do not recommend limiting workaround with stride 1 only. @daniellowell what is your opinion?

Comment (daniellowell) [reply src/solver/conv_multipass_wino3x3WrW.cpp:386]: All fp16. I haven’t narrowed it down to stride 1.

File: src/solver/conv_multipass_wino3x3WrW.cpp [hunk 2]
```diff
@@ -381,15 +381,33 @@ bool ConvWinograd3x3MultipassWrW<WinoDataH, WinoFilterH, WinoDataW, WinoFilterW>
...
-        if(miopen::IsDisabled(MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_F3X4{}))
-            return false;
-    if(wino_data_tile == 3 && wino_filter_tile == 5)
-        if(miopen::IsDisabled(MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_F3X5{}))
-            return false;
-    if(wino_data_tile == 3 && wino_filter_tile == 6)
-        if(miopen::IsDisabled(MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_F3X6{}))
-            return false;
+
+    const std::string name = params.GetStream().GetDeviceName();
```
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:385]: Please add
```C++
#define WORKAROUND_SWDEV_234193 1
````
at the beginning of file and put the workaround code within `#if WORKAROUND_SWDEV_234193... #endif`.

File: src/solver/conv_multipass_wino3x3WrW.cpp (hunk 1 above)
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:386]: [Recommendation]
```suggestion
    if(params.IsFp16() && StartsWith(name, "gfx908"))
```
@daniellowell what is your opinion?

Comment (shurale-nkn) [reply src/solver/conv_multipass_wino3x3WrW.cpp:386]: SWDEV_234193 does not contain conv3x3 stride2, only stride 1

Comment (daniellowell) [reply src/solver/conv_multipass_wino3x3WrW.cpp:386]: Ok, but are these solvers safe for stride 2?

Comment (shurale-nkn) [reply src/solver/conv_multipass_wino3x3WrW.cpp:386]: It can pass verification steps. @Kirpich30000 ?

Comment (atamazov) [reply src/solver/conv_multipass_wino3x3WrW.cpp:386]: Stride 1 solvers are passing verification as well, however we see the problems. Perhaps the SWDEV is not related to these solvers at all. Well, it _is related_, but _the reason_ could be different. For example, by disabling these solvers we are freeing some workspace, thus allowing some other workspace-greedy solver to run _instead_ of some other solver, which needs less workspace but is really faulty.

We need complete logs collected at level 5 for analysis.

As far as we do not have information about the real cause of the issue, let's act as simple as possible, but with a guaranteed effect.

Comment (shurale-nkn) [reply src/solver/conv_multipass_wino3x3WrW.cpp]: better safe than sorry. ok

Comment (shurale-nkn) [reply src/solver/conv_multipass_wino3x3WrW.cpp]: And we have several log files in the ticket.

**Comment by Kirpich30000** [comment]
Please note that this PR will have perf impact. As noted in internal email there is a chance to regain some of the loss. Also I'm not sure about filtering `gfx908` only because depending on the exact root cause it could be inappropriate.

**Comment by daniellowell** [comment]
> Please note that this PR will have perf impact. As noted in internal email there is a chance to regain some of the loss. Also I'm not sure about filtering `gfx908` only because depending on the exact root cause it could be inappropriate.

Noted. Performance impact for the release should not be an issue on gfx908. Let's see if we can do anything about the instability (if anything) in the mean time. Since we are only seeing it on gfx908, let's be somewhat conservative about disabling it for other ASICs.

**Comment by atamazov** [comment]
@daniellowell 
> Let's see if we can do anything about the instability (if anything) in the mean time.

Yes. We still do not know who is guilty -- MP Winograd or iGemm.

As I said [here](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/252#discussion_r432886893), I can try to look into logs collected at level 5.

@shurale-nkn 
> And we have several log files in the ticket.

I need two logs collected in a row on the same system, with the same docker and the same build of MIOpen. One with default settings, another one with three MP Winograd solvers disabled via environment.

Or we can quickly implement variables for iGemm solvers and try disabling them individually.

Then we'll have the list of candidates (three MP Winograds + some iGemms) and will be able to perform some rigorous correctness testing.

File: src/solver/conv_multipass_wino3x3WrW.cpp [hunk 3]
```diff
@@ -35,9 +35,9 @@
 #include <miopen/generic_search.hpp>
 #include <miopen/tensor.hpp>
 #include <miopen/solver.hpp>
-
 #if(MIOPEN_BACKEND_HIP && MIOPEN_USE_ROCBLAS)
 #define WORKAROUND_SWDEV_203031 1 // See also issues #2075, #2067
+#define WORKAROUND_SWDEV_234193 1
```
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:40]: [anchor]

**Pull request merged**